### PR TITLE
EZP-29104: Impl. ImageAsset field type

### PR DESCRIPTION
**Pull request by adamwojs**
> JIRA: https://jira.ez.no/browse/EZP-29104
> Depends on https://github.com/ezsystems/ezpublish-kernel/pull/2403

## Description 

Impl. of `\EzSystems\RepositoryForms\FieldType\FieldDefinitionFormMapperInterface` and `\EzSystems\RepositoryForms\FieldType\FieldValueFormMapperInterface` for Image Asset Field Type.

### CODE REVIEW

File: lib/FieldType/DataTransformer/ImageAssetValueTransformer.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+use eZ\Publish\Core\FieldType\ImageAsset\Value;
+use Symfony\Component\Form\DataTransformerInterface;
+
+class ImageAssetValueTransformer extends AbstractBinaryBaseTransformer implements DataTransformerInterface
+{
+    public function transform($value)
+    {
+        if (!$value instanceof Value) {
+            return null;
```
Comment (webhdx) [inline lib/FieldType/DataTransformer/ImageAssetValueTransformer.php]: To make it more future proof I'd rather return `null` for `$value = null` but for `!$value instanceof Value` it should throw `\Symfony\Component\Form\Exception\TransformationFailedException`.

File: lib/FieldType/Mapper/ImageAssetFormMapper.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+{
+    /** @var \eZ\Publish\API\Repository\FieldTypeService */
+    private $fieldTypeService;
+
+    public function __construct(FieldTypeService $fieldTypeService)
+    {
+        $this->fieldTypeService = $fieldTypeService;
+    }
+
+    public function mapFieldDefinitionForm(FormInterface $fieldDefinitionForm, FieldDefinitionData $data)
```
Comment (webhdx) [inline lib/FieldType/Mapper/ImageAssetFormMapper.php]: I don't think you need to implement `FieldDefinitionFormMapperInterface` and define empty method. It should work without these.

File: lib/Form/Type/FieldType/ImageAssetFieldType.php [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+use Symfony\Component\Form\Extension\Core\Type\HiddenType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\Form\FormView;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class ImageAssetFieldType extends AbstractType
+{
+    /** @var \eZ\Publish\API\Repository\ContentService */
+    private $contentService;
```
Comment (webhdx) [inline lib/Form/Type/FieldType/ImageAssetFieldType.php]: Add empty lines between properties.

File: lib/Form/Type/FieldType/ImageAssetFieldType.php [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    public function getParent()
+    {
+        return BinaryBaseFieldType::class;
+    }
+
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder
+            ->add(
+                'destinationContentId',
```
Comment (webhdx) [inline lib/Form/Type/FieldType/ImageAssetFieldType.php]: Nitpick: no need to make it multiline

File: lib/FieldType/DataTransformer/ImageAssetValueTransformer.php [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+declare(strict_types=1);
+
+namespace EzSystems\RepositoryForms\FieldType\DataTransformer;
+
+use eZ\Publish\Core\FieldType\ImageAsset\Value;
+use Symfony\Component\Form\DataTransformerInterface;
+
+class ImageAssetValueTransformer extends AbstractBinaryBaseTransformer implements DataTransformerInterface
+{
+    public function transform($value)
```
Comment (mikadamczyk) [inline lib/FieldType/DataTransformer/ImageAssetValueTransformer.php]: what about docblock and return type hint here and `reverseTransform()`?

File: lib/FieldType/Mapper/ImageAssetFormMapper.php [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+use EzSystems\RepositoryForms\Form\Type\FieldType\ImageAssetFieldType;
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class ImageAssetFormMapper implements FieldDefinitionFormMapperInterface, FieldValueFormMapperInterface
+{
+    /** @var \eZ\Publish\API\Repository\FieldTypeService */
+    private $fieldTypeService;
+
+    public function __construct(FieldTypeService $fieldTypeService)
```
Comment (mikadamczyk) [inline lib/FieldType/Mapper/ImageAssetFormMapper.php]: docblocks?

**Comment by adamwojs** [comment]
PR updated according to @webhdx @mikadamczyk suggestions.

**Comment by adamwojs** [comment]
PR ready for the final code review @mikadamczyk @webhdx

File: lib/Data/FieldType/SelectionOption.php [hunk 7]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+/**
+ * This file is part of the eZ Publish Kernel package.
```
Comment (mikadamczyk) [inline lib/Data/FieldType/SelectionOption.php]: This is not a part of the Kernel package. We can add `declare(strict_types=1);`

File: lib/Data/FieldType/SelectionOption.php [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace EzSystems\RepositoryForms\Data\FieldType;
+
+use eZ\Publish\API\Repository\Values\ValueObject;
+
+class SelectionOption extends ValueObject
+{
+    public $isDefault;
```
Comment (mikadamczyk) [inline lib/Data/FieldType/SelectionOption.php]: Missing dockblock and an empty line between properties

File: bundle/Resources/config/fieldtypes.yml [hunk 9]
```diff
@@ -163,3 +163,8 @@ services:
             - { name: ez.fieldFormMapper.value, fieldType: ezemail }
         calls:
             - [setFormType, ['Symfony\Component\Form\Extension\Core\Type\EmailType']]
+
+    EzSystems\RepositoryForms\FieldType\Mapper\ImageAssetFormMapper:
+        tags:
+            - { name: ez.fieldFormMapper.definition, fieldType: ezimageasset }
```
Comment (webhdx) [inline bundle/Resources/config/fieldtypes.yml]: `ez.fieldFormMapper.definition` is not needed as the class is not implementing `FieldDefinitionFormMapperInterface`

**Pull request merged**